src/snmp_agent.c

```diff
@@ -0,0 +1,1513 @@
+    snmp_agent_free_data(&dd);
+  }
+
+  llist_destroy((*td)->columns);
+
+  void *key = NULL;
+  void *value = NULL;
+
+  c_avl_destroy((*td)->index_instance);
+  (*td)->index_instance = NULL;
+
+  while (c_avl_pick((*td)->instance_index, &key, &value) == 0) {
+    sfree(key);
+    sfree(value);
+  }
+  c_avl_destroy((*td)->instance_index);
+  (*td)->instance_index = NULL;
+
+  sfree((*td)->name);
+  sfree(*td);
+
+  return;
+}
+
+static int
+snmp_agent_table_oid_handler(struct netsnmp_mib_handler_s *handler,
+                             struct netsnmp_handler_registration_s *reginfo,
+                             struct netsnmp_agent_request_info_s *reqinfo,
+                             struct netsnmp_request_info_s *requests) {
+
+  if (reqinfo->mode != MODE_GET && reqinfo->mode != MODE_GETNEXT) {
+    DEBUG(PLUGIN_NAME ": Not supported request mode (%d).", reqinfo->mode);
+    return SNMP_ERR_NOERROR;
+  }
+
+  oid_t oid;
+  char temp[DATA_MAX_NAME_LEN];
+
+  pthread_mutex_lock(&g_agent->lock);
+
+  memcpy(oid.oid, requests->requestvb->name,
+         sizeof(oid.oid[0]) * requests->requestvb->name_length);
+  oid.oid_len = requests->requestvb->name_length;
+
+  snmp_agent_oid_to_string(temp, sizeof(temp), &oid);
+
+  DEBUG(PLUGIN_NAME ": Get request received for table OID '%s'.", temp);
+
+  for (llentry_t *te = llist_head(g_agent->tables); te != NULL; te = te->next) {
+    table_definition_t *td = (table_definition_t *)te->value;
+
+    for (llentry_t *de = llist_head(td->columns); de != NULL; de = de->next) {
+      data_definition_t *dd = (data_definition_t *)de->value;
+
+      for (int i = 0; i < dd->oids_len; i++) {
+
+        int ret = snmp_oid_ncompare(oid.oid, oid.oid_len,
+                                dd->oids[i].oid, dd->oids[i].oid_len,
+                                MIN(oid.oid_len, dd->oids[i].oid_len));
+        if (ret != 0) continue;
+
+        if (td->index_oid.oid_len) {
+          int index = oid.oid[oid.oid_len - 1];
+          char *instance;
+
+          ret = c_avl_get(td->index_instance, &index, (void **)&instance);
+          if (ret != 0) {
+            DEBUG(PLUGIN_NAME ": Nonexisting index '%d' requested.", index);
+            pthread_mutex_unlock(&g_agent->lock);
+            return SNMP_NOSUCHINSTANCE;
+          }
+
+          if (dd->is_instance) {
+
+            requests->requestvb->type = ASN_OCTET_STR;
+            snmp_set_var_typed_value(requests->requestvb,
+                                     requests->requestvb->type, instance,
+                                     strlen((instance)));
+
+            pthread_mutex_unlock(&g_agent->lock);
+
+            return SNMP_ERR_NOERROR;
+          }
+
+          format_name(temp, sizeof(temp), hostname_g, dd->plugin, instance,
+                      dd->type, dd->type_instance);
+          DEBUG(PLUGIN_NAME ": Identifier '%s'", temp);
+
+          value_t *values;
+          size_t values_num;
```

Comment:
Please initialize `values` and `values_num` to `NULL` / zero so that it's obvious that they are never used uninitialized.